src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseStream.cs

```diff
@@ -131,21 +131,57 @@ public override Task<int> ReadAsync(byte[] buffer, int offset, int count, Cancel
             _state.TcsReadFromResponseStream =
                 new TaskCompletionSource<int>(TaskCreationOptions.RunContinuationsAsynchronously);
 
+            _state.TcsQueryDataAvailable =
+                new TaskCompletionSource<int>(TaskCreationOptions.RunContinuationsAsynchronously);
+            _state.TcsQueryDataAvailable.Task.ContinueWith((previousTask) => 
+                {
+                    if (previousTask.IsFaulted)
+                    {
+                        _state.TcsReadFromResponseStream.TrySetException(previousTask.Exception.InnerException);
+                    }
+                    else if (previousTask.IsCanceled)
+                    {
+                        _state.TcsReadFromResponseStream.TrySetCanceled();
+                    }
+                    else
+                    {
+                        int bytesToRead;
+                        int bytesAvailable = previousTask.Result;
+                        if (bytesAvailable > count)
+                        {
+                            bytesToRead = count;
+                        }
+                        else
+                        {
+                            bytesToRead = bytesAvailable;
+                        }
+                        
+                        lock (_state.Lock)
+                        {
+                            if (!Interop.WinHttp.WinHttpReadData(
+                                _state.RequestHandle,
+                                Marshal.UnsafeAddrOfPinnedArrayElement(buffer, offset),
+                                (uint)bytesToRead,
+                                IntPtr.Zero))
+                            {
+                                _state.TcsReadFromResponseStream.TrySetException(
+                                    new IOException(SR.net_http_io_read, WinHttpException.CreateExceptionUsingLastError()));
+                            }
+                        }
+                    }
+                });
```

Comment:
This should include:
```C#
), CancellationToken.None, TaskContinuationOptions.None, TaskScheduler.Default);
```
the important piece being the TaskScheduler.Default.  Later you could decide to also pass in a value for the object state arg, which would potentially let you eliminate the closure and delegate allocations per call.